src/main.rs

```diff
@@ -122,15 +195,33 @@ fn process_solve(theory: &Theory, bound: Option<BoundCommand>, count: Option<i32
         None
     };
 
-    strategy.add(Model::new(), selector);
     let mut found = 0;
 
-    while !strategy.empty() {
-        if count.is_some() && found >= count.unwrap() {
-            break;
+    // FIXME: use dynamic dispatching
+    match strategy {
```

Comment:
@KiChjang I particularly want your opinion here. Do you recommend using trait objects?

Reply:
If you have a closed set of variants that you can iterate over with, then I would suggest sticking with enum. What problem are you trying to solve now that made you think of using trait objects?

Reply:
Something that I am exploring in this project is to test different strategies for implementing the model-finding algorithm, things like: using a FIFO or LIFO strategy for exploring search branches, various strategies to bound the computation, different ways of choosing the next formula (sequent) to expand a solution, different solutions for equational reasoning, etc.
Each of these decision points have their own advantages and disadvantages and I want the user to be able to choose the particular combination of strategies (and possibly even create a DSL so the user can change the strategy in the middle of the model-finding).
Currently, each variation in the solution is a trait that can be implemented by different objects (like FIFO and LIFO for choosing the next branch). So for a run of the algorithm, there is a combinatorial number of arrangements of these strategies. Even though they can be enumerated, there are going to be a lot of them. So wondering if trait objects are the way to go.

Reply:
Hmm... so looking further into what you're doing with this match, you can in fact define a new method on your enum called `solve`, so that instead of performing the algorithm for your strategy here, you'd put the logic inside of this new `solve` method.

Alternatively, you can indeed create a new trait called `Strategy` that has a function called `solve` and have both `FIFO` and `LIFO` implement it. Then you can make `process_solve` generic like this: `fn process_solve<S: Strategy>(theory: &Theory, bound: Option<BoundCommand>, strategy: S, count: Option<i32>, color: bool)`, eliminating the need for dynamic dispatch and trait objects.

Reply:
But wouldn't that bring me back exactly where I am right now? currently, `FIFO` and `LIFO` both implement a trait  `StrategyTrait`. The problem is when I want to pass an object `strategy` to `solve` (which takes `S: StrategyTrait`), the concrete type of `strategy` must be known at compile time. So I can't change the value of my `strategy` based on the user input (unless I use a trait object?).

Reply:
I see what you're trying to do here, and I guess what I wanted to ask is, is it possible to make it so that you only have an enum called `Strategy` that looks something like the following:
```rust
pub enum Strategy<'s, S: 's + SequentTrait, M: ModelTrait, Sel: SelectorTrait<Item=&'s S>> {
    Fifo(VecDeque<(M, Sel)>),
    Lifo(VecDeque<(M, Sel)>),
}
```

Reply:
@KiChjang damn, that's cool! I don't see any reason for it not to work but I guess my question is why is this better than using trait objects? or in general, what are the pros and cons?

Reply:
Ultimately the reason here is that you control exactly how many kinds of strategy that you can have, and so the commonality between your variants isn't just functions; in fact, thinking more about it, you could also model it like so:
```rust
pub enum StrategyKind { Fifo, Lifo }

pub struct Strategy<'s, S: 's + SequentTrait, M: ModelTrait, Sel: SelectorTrait<Item=&'s S>>{
    queue: VecDeque<(M, Sel)>,
    kind: StrategyKind,
}
```
The downside of this is that you must ensure all strategies have the same VecDeque data structure, which may not be what you want.

In general, the trade-off that you're making here is data uniformity vs performance -- traits are very loose on the contract between data and behavior, but you must then pay in performance for static dispatch or dynamic dispatch; on the other hand, using a struct and calling its methods is basically just a function call, the cost of it is that you must ensure all kinds of strategies share the same shape of the data. Enums are kind of in-between -- you definitely don't pay as much as traits during runtime, but you'll always have to match in your methods on the enum, i.e. there is always a tag lookup during runtime, but in return you get to pick the fields for each kind of your strategy.

Since I don't exactly know what your future strategies are going to look like, I'd go with using the enum-with-fields approach.

Reply:
I decided to go with a hybrid approach, keeping `FIFO` and `LIFO` as they are. But I also added a `Dispatch` enum that implements `StrategyTrait` and the two cases of `Dispatch` hold instances of FIFO and LIFO.